sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -528,30 +549,40 @@ private class BulkWriter(container: CosmosAsyncContainer,
         try {
           val itemOperation = resp.getOperation
           val itemOperationFound = activeBulkWriteOperations.remove(itemOperation)
-          assume(itemOperationFound) // can't find the item operation in list of active operations!
-          val context = itemOperation.getContext[OperationContext]
-          val itemResponse = resp.getResponse
-
-          if (resp.getException != null) {
-            Option(resp.getException) match {
-              case Some(cosmosException: CosmosException) =>
-                handleNonSuccessfulStatusCode(
-                  context, itemOperation, itemResponse, isGettingRetried, Some(cosmosException))
-              case _ =>
-                log.logWarning(
-                  s"unexpected failure: itemId=[${context.itemId}], partitionKeyValue=[" +
-                    s"${context.partitionKeyValue}], encountered , attemptNumber=${context.attemptNumber}, " +
-                    s"exceptionMessage=${resp.getException.getMessage}, " +
-                  s"Context: ${operationContext.toString} $getThreadInfo", resp.getException)
-                captureIfFirstFailure(resp.getException)
-                cancelWork()
+          if (!itemOperationFound) {
+            // can't find the item operation in list of active operations!
+            log.logInfo(s"Cannot find active operation for '${itemOperation.getOperationType} " +
+            s"${itemOperation.getPartitionKeyValue}/${itemOperation.getId}'. This can happen when " +
+            s"retries get re-enqueued.")
+
+            if (pendingBulkWriteRetries.remove(itemOperation)) {
```

Comment:
After thinking more , I think  we probably should only remove from the pendingBulkWriteRetries if it is a successful response. 

There are few situations can happen with current change:
1.  Both requests succeeded -> works fine
2. First request succeeded, second request failed -> works fine 
3. First request failed and scheduled retry,  second request succeeded -> works fine
4. **First request failed and pending retry, second request succeeded -> Will have issue**
5. **First request failed and scheduled retry, second request failed -> will have issue**
6. First request failed and pending retry, second request failed -> will work fine

Reply:
I think 5) would have been ok - but 4) definitely would have caused issues. Changed this now to handle the response if we either were able to remove the operation form activeOperations or pendingRetries. Pending Retry would also be removed in both cases - but if response has retriable failure we would add another retry